File: leetcode/Algorithms/160.IntersectionOfTwoLinkedLists/Solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */

public class Solution {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        HashMap<Integer, Integer> hashMap = new HashMap<>();

        while (headA != null) {
            hashMap.put(headA.val, 0);
            headA = headA.next;
        }

        while (headB != null && !hashMap.containsKey(headB.val)) {
            headB = headB.next;
        }

        return headB;
    }
}

File: leetcode/Algorithms/160.IntersectionOfTwoLinkedLists/Solution.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */

struct ListNode *getIntersectionNode(struct ListNode *headA, struct ListNode *headB) {
    if (headA == NULL || headB == NULL) return NULL;

    struct ListNode *tmp_a = headA, *tmp_b = headB;
    while (tmp_a != tmp_b) {
        tmp_a = tmp_a == NULL ? headB : tmp_a->next;
        tmp_b = tmp_b == NULL ? headA : tmp_b->next;
    }

    return tmp_a;
}
